shell/platform/linux/fl_value.cc

```diff
@@ -92,6 +92,120 @@ static ssize_t fl_value_lookup_index(FlValue* self, FlValue* key) {
   return -1;
 }
 
+static void int_to_string(GString* buffer, int64_t value) {
+  g_string_append_printf(buffer, "%" G_GINT64_FORMAT, value);
+}
+
+static void float_to_string(GString* buffer, double value) {
+  g_string_append_printf(buffer, "%.16f", value);
+
+  // Strip trailing zeros
+  int zero_count = 0;
+  for (int i = buffer->len - 1; i >= 0; i--) {
+    // Leave one zero after a decimal point
+    if (buffer->str[i] == '.') {
+      zero_count = zero_count == 0 ? 0 : zero_count - 1;
+      break;
+    }
+    if (buffer->str[i] != '0')
+      break;
+    zero_count++;
+  }
+  g_string_truncate(buffer, buffer->len - zero_count);
+}
+
+static void value_to_string(GString* buffer, FlValue* value) {
+  switch (value->type) {
+    case FL_VALUE_TYPE_NULL:
+      g_string_append(buffer, "null");
+      return;
+    case FL_VALUE_TYPE_BOOL:
+      if (fl_value_get_bool(value))
+        g_string_append(buffer, "true");
+      else
+        g_string_append(buffer, "false");
+      return;
+    case FL_VALUE_TYPE_INT:
+      int_to_string(buffer, fl_value_get_int(value));
+      return;
+    case FL_VALUE_TYPE_FLOAT:
+      float_to_string(buffer, fl_value_get_float(value));
+      return;
+    case FL_VALUE_TYPE_STRING: {
+      g_string_append(buffer, fl_value_get_string(value));
```

Comment:
This should output quotation marks, so that numbers and numbers-as-strings are distinguishable.

Edit: I see later that you said that you are specifically matching the output of Dart's toString, which doesn't do that. I think it's pretty unfortunate that `print([42, null, 'null', '42'].toString());` prints `[42, null, null, 42]` in Dart, and in the context of debugging type mistakes `FlValue` structures that could be a non-trivial problem. But I'll leave it up to you since I can see the argument for matching Dart.

Reply:
I was very surprised that Dart doesn't add the quotation marks (I originally had them in `fl_value_to_string`). I decided that matching Dart was the best thing to do, as you're likely to be printing out things that were generated in Dart code.